stubs/python-decouple/decouple.pyi

```diff
@@ -0,0 +1,98 @@
+from typing import Any, Generic, TextIO, TypeVar, overload
+
+_T = TypeVar("_T")
+_TCsv = TypeVar("_TCsv", bound=Sequence[Any])
+
+PYVERSION: Incomplete  # undocumented
+DEFAULT_ENCODING: str  # undocumented
+TRUE_VALUES: set[str]  # undocumented
+FALSE_VALUES: set[str]  # undocumented
+
+def read_config(parser: ConfigParser, file: TextIO) -> None: ...  # undocumented
+def strtobool(value: str | bool) -> bool: ...  # undocumented
+
+class UndefinedValueError(Exception): ...
+class Undefined: ...
+
+undefined: Undefined
+
+class Config:
+    repository: RepositoryEmpty
+    def __init__(self, repository: RepositoryEmpty) -> None: ...
+    @overload
+    def get(self, option: str, default: str = ...) -> str: ...
+    @overload
+    def get(self, option: str, cast: Callable[..., _T]) -> _T: ...
+    @overload
+    def get(self, option: str, default: str, cast: Callable[..., _T]) -> _T: ...
+    @overload
+    def __call__(self, option: str, default: str = ...) -> str: ...
+    @overload
+    def __call__(self, option: str, cast: Callable[..., _T]) -> _T: ...
+    @overload
+    def __call__(self, option: str, default: str, cast: Callable[..., _T]) -> _T: ...
+
+class RepositoryEmpty:  # undocumented
+    def __init__(self, source: str = ..., encoding: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositoryIni(RepositoryEmpty):
+    SECTION: str
+    parser: ConfigParser
+    def __init__(self, source: str, encoding: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositoryEnv(RepositoryEmpty):
+    data: dict[str, str]
+    def __init__(self, source: str, encoding: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositorySecret(RepositoryEmpty):
+    data: dict[str, str]
+    def __init__(self, source: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class AutoConfig:
+    SUPPORTED: OrderedDict[str, RepositoryEmpty]
+    encoding: str
+    search_path: str | None
+    config: Config | None
+    def __init__(self, search_path: str | None = None) -> None: ...
+    @overload
+    def __call__(self, option: str, default: str = ...) -> str: ...
+    @overload
+    def __call__(self, option: str, cast: Callable[..., _T]) -> _T: ...
+    @overload
+    def __call__(self, option: str, default: str, cast: Callable[..., _T]) -> _T: ...
+
+config: AutoConfig
+
+class Csv(Generic[_T, _TCsv]):
+    cast: Callable[..., _T]
+    delimiter: str
+    strip: str
+    post_process: Callable[..., _TCsv]
+    def __call__(self, value: str) -> _TCsv: ...
+    def __init__(
+        self, cast: Callable[..., _T] = str, delimiter: str = ..., strip: str = ..., post_process: Callable[..., _TCsv] = list[_T]
+    ) -> None: ...
```

Comment:
```suggestion
    @overload
    def __init__(
        self: Csv[str, list[str]], cast: Callable[[str], str] = str, delimiter: str = ..., strip: str = ..., post_process: Callable[[Iterable[str]], list[str]] = list
    ) -> None: ...
    @overload
    def __init__(
        self: Csv[_T, list[_T]], cast: Callable[[str], _T], delimiter: str = ..., strip: str = ..., post_process: Callable[[Iterable[_T]], list[_T]] = list
    ) -> None: ...
    ) -> None: ...
    @overload
    def __init__(
        self: Csv[str, _TCsv], cast: Callable[[str], _T] = str, delimiter: str = ..., strip: str = ..., *, post_process: Callable[Iterable[str], _TCsv]
    ) -> None: ...
    @overload
    def __init__(
        self: Csv[_T, _TCsv], cast: Callable[[str], _T], delimiter: str, strip: str, post_process: Callable[[Iterable[_T]], _TCsv]
```
I think this should be something along these lines. The arguments of `cast`/`post_process` are very much defined by the implementation and since they are optional parameters you will need to provide overloads, because otherwise _T/_TCsv will not get bound.

It's also worth noting that you will need to import `Iterable` to make this suggestion work.

Reply:
This is a great suggestion!  I didn't think about setting the type on `self` to influence the generic type.

I updated the PR with your suggestions, everything looks great, but there might be a weird regression with cast by itself. For some reason, mypy always thinks the return is a `str`.

Example:
```python
from decouple import Choices, Csv
from typing_extensions import reveal_type

reveal_type(Choices(flat=[1, 2])(""))
reveal_type(Choices(choices=[(1, "one"), (2, "two")])(""))
reveal_type(Choices(cast=int, flat=[1, 2])(""))
reveal_type(Choices(cast=int, choices=[(1, "one"), (2, "two")])(""))
reveal_type(Choices(flat=[1, 2], choices=[(1, "one"), (2, "two")])(""))
reveal_type(Choices(cast=int, flat=[1, 2], choices=[(1, "one"), (2, "two")])(""))
reveal_type(Choices(cast=int)(""))
```

Output:
```
test.py:4: note: Revealed type is "builtins.int"
test.py:5: note: Revealed type is "builtins.int"
test.py:6: note: Revealed type is "builtins.int"
test.py:7: note: Revealed type is "builtins.int"
test.py:8: note: Revealed type is "builtins.int"
test.py:9: note: Revealed type is "builtins.int"
test.py:10: note: Revealed type is "builtins.str"
test.py:10: error: Argument "cast" to "Choices" has incompatible type "type[int]"; expected "Callable[[str], str]"  [arg-type]
```

Reply:
Would you know how to address the `reveal_type(Choices(cast=int)(""))` case?

Reply:
I am sorry, I messed this up. Let me delete my comments and update it.

Reply:
Thank you for this suggestion! I did not realize you could suggest the types of `_T` and `_Csv` through the `self` argument.  Really interesting stuff.

After I implemented your suggestions, I am getting some regressions on my code while using this type file.  Would you know how to fix it?  Also, I am wondering if the `Iterable` is having some  unwanted side effects.  Also, the errors I am seeing are coming from mypy, I haven't checked the other type checkers.

```python
from decouple import Choices, Csv
from typing_extensions import reveal_type

# No error and correct type
reveal_type(Csv()(""))  # note: Revealed type is "builtins.list[builtins.str]"

# No error and correct type
reveal_type(Csv(cast=int)(""))  # note: Revealed type is "builtins.list[builtins.int]"

# error: Argument "post_process" to "Csv" has incompatible type "type[set[Any]]"; expected "Callable[[Iterable[str]], list[str]]"  [arg-type]
reveal_type(Csv(post_process=set)(""))  # note: Revealed type is "builtins.list[builtins.str]"

# error: Argument 1 to "Csv" has incompatible type "type[int]"; expected "Callable[[str], str]"  [arg-type]
# error: Argument 4 to "Csv" has incompatible type "type[tuple[Any, ...]]"; expected "Callable[[Iterable[str]], list[str]]"  [arg-type]#
reveal_type(Csv(int, ",", " ", tuple)(""))

# error: Argument "cast" to "Csv" has incompatible type "type[int]"; expected "Callable[[str], str]"  [arg-type]
# error: Argument "post_process" to "Csv" has incompatible type "type[tuple[Any, ...]]"; expected "Callable[[Iterable[str]], list[str]]"  [arg-type]
reveal_type(Csv(cast=int, post_process=tuple)(""))

# error: Argument "cast" to "Csv" has incompatible type "type[int]"; expected "Callable[[str], str]"  [arg-type]
# error: Argument "post_process" to "Csv" has incompatible type "type[tuple[Any, ...]]"; expected "Callable[[Iterable[str]], list[str]]"  [arg-type]
reveal_type(Csv(cast=int, delimiter=",", post_process=tuple)(""))

```